### PR TITLE
UI for setting the reason why SHF is waiting on an applicant

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story:  https://www.pivotaltracker.com/story/show/143810729

This is the 2nd PR (the 2nd half) for the story.

Changes proposed in this pull request:
1.  Seed the db with 1 reason: the "other/custom" reason because it is used specifically.  (Other reasons can be entered via the usual web pages/interface).
2. UI: the _reasons for waiting_ has been moved to above the _application status buttons
3.  Uses AJAX to update the db immediately when a reason is selected from the list.  The db is also updated immediately when the text for a 'custom/other' reason is entered.

### Screenshots (Optional):

reasonwaiting-select-reason-sv

reasonwaiting-select-reason-en

---

#### The reason isn't shown if the status is not 'Waiting for Applicant'
reasonwaiting-hidden-unless-status-en

---

#### The text field for typing in the "other reason" only shows if that reason is selected

reasonwaiting-other-reason-text-en

Ready for review:
@patmbolger @thesuss @RobertCram

### CODE REVIEW

File: app/models/admin_only/member_app_waiting_reason.rb [hunk 1]
```diff
@@ -12,15 +12,25 @@ def other_reason_placeholder?
 
 
     # This is effectively a CONSTANT that is used in the UI
-    def self.other_reason_name
-      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason')
+    def self.other_reason_name(locale = I18n.locale)
+      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason', locale)
```
Comment (RobertCram) [inline app/models/admin_only/member_app_waiting_reason.rb:16]: The second parameter to the translate method should be an options hash. As it stands now the locale is ignored, and the method always returns the swedish translation

I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason', locale: locale)

File: app/models/admin_only/member_app_waiting_reason.rb [hunk 2]
```diff
@@ -12,15 +12,25 @@ def other_reason_placeholder?
...
     end
 
 
+    def self.other_reason_desc(locale = I18n.locale)
+      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason_desc', locale)
+    end
```
Comment (RobertCram) [inline app/models/admin_only/member_app_waiting_reason.rb:22]: Same as above.

**Comment by RobertCram** [comment]
Great work, Ashley. Apart from my inline comment, two minor remarks:

1. There should probably be some sort of protection against changing or deleting the "other/custom reason" record?

2. In the UI, I would personally prefer the "Other (enter the reason)" to be the last option in the list

**Comment by weedySeaDragon** [comment]
I know the build fails.  I can't figure out why the cucumber feature tests are failing.   Any help is appreciated!

fyi:  the feature that fail is this file: `features/membership-application-status/admin-custom-reason-waiting.feature`

**Comment by weedySeaDragon** [comment]
passing locally on my machine with selenium, but not here with semaphore. <sigh>

**Comment by weedySeaDragon** [comment]
Closing this. Will replace it with with a much cleaner approach that @patmbolger came up with. 👍 

We sure did learn a lot about poltergeist, phantom, selenium, and `data-remote` (oh my!).  
 Hopefully we'll write up / document that.  Meanwhile, where's a short list:

### Lessons learned:
- phantom.js will silently fail if you use `let` in your Javascript (ECMS 6.0).  (Use 'var' instead)  Ref: [poltergeist issue 823: use of JavaScript's let will fail silently ](https://github.com/teampoltergeist/poltergeist/issues/823)
- waiting for a  "Then I should not see"  used `expect(page).not_to have_content content`.  However, this does _not_ make Capybara wait for ajax, which leads to all kind of messes (rails still running while the test is finished, etc.). (This made some of the tests fail due to timeout because they were waiting forever.)
  use Capybara's `assert_text` instead

- use `data-remote` to send info back to the controller, even if the parameters aren't what is 'normally' expected by the Controller.  Let the controller parse and deal with whatever parameters are sent.  Dealing with them in your Javascript is ugly and harder to maintain.

**Comment by patmbolger** [comment]
Good list! 

In the 2nd item, the `assert_` Capybara methods do not wait on outstanding AJAX requests to complete, but they _do_ wait as long as the Capybara timeout setting (which I think is 10 seconds by default).  This is usually a much better approach than using Rspec `expect`.

A 4th item is the mechanism that @RobertCram developed to tell Capybara to wait until outstanding AJAX requests have all completed - that will prove to be useful, I am sure.  We should make that a separate test step altogether, so we can use it as needed (e.g. `Then I wait until AJAX requests complete`).  (kind of wonky, but since we don't expose the cute tests to the client I suppose we can get away with it).

**Pull request closed without merging**